x/lockup/keeper/lock.go

```diff
@@ -722,3 +722,78 @@ func (k Keeper) getCoinsFromLocks(locks []types.PeriodLock) sdk.Coins {
 	}
 	return coins
 }
+
+func (k Keeper) RebondTokens(ctx sdk.Context, lockID uint64, owner sdk.AccAddress, coins sdk.Coins) error {
+	lock, err := k.GetLockByID(ctx, lockID)
+	if err != nil {
+		return err
+	}
+
+	if lock.Owner != owner.String() {
+		return fmt.Errorf("lock %d is not owned by %s", lockID, owner)
+	}
+
+	if !lock.IsUnlocking() {
+		return fmt.Errorf("lock %d is not unlocking, rebonding only possible in unlocking stage", lockID)
+	}
+
+	// If all checks pass, we can rebond the tokens
+	err = k.rebondTokens(ctx, owner, *lock, coins)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// rebondTokens rebonds the specified amount of coins from the lock. If nil, rebonds all tokens.
+func (k Keeper) rebondTokens(ctx sdk.Context, owner sdk.AccAddress, lock types.PeriodLock, coins sdk.Coins) error {
+	var rebondedLock types.PeriodLock
+	var err error
+	if coins != nil && !coins.IsEqual(lock.Coins) {
```

Comment:
coins.IsZero won't return true if the coins is nil, I think it will panic, because it tries to reference internal `sdk.Coin` field, which would be non-existent if coins is nil